sys/net/network_layer/ng_ndp/node/ng_ndp_node.c

```diff
@@ -55,139 +55,110 @@ static ng_pktqueue_t *_alloc_pkt_node(ng_pktsnip_t *pkt)
     return NULL;
 }
 
-kernel_pid_t ng_ndp_node_next_hop_l2addr(uint8_t *l2addr, uint8_t *l2addr_len,
-                                         kernel_pid_t iface, ipv6_addr_t *dst,
-                                         ng_pktsnip_t *pkt)
+bool ng_ndp_node_next_hop_ipv6_addr(ipv6_addr_t *next_hop_ip,
+                                    kernel_pid_t *iface,
+                                    ipv6_addr_t *dst)
 {
-    ipv6_addr_t *next_hop_ip = NULL, *prefix = NULL;
-
 #ifdef MODULE_NG_IPV6_EXT_RH
     next_hop_ip = ng_ipv6_ext_rh_next_hop(hdr);
+    if (next_hop_ip != NULL) {
+        return true;
+    }
 #endif
 #ifdef MODULE_FIB
     size_t next_hop_size = sizeof(ipv6_addr_t);
     uint32_t next_hop_flags = 0;
-    ipv6_addr_t next_hop_actual;    /* FIB copies address into this variable */
-
-    if ((next_hop_ip == NULL) &&
-        (fib_get_next_hop(&iface, next_hop_actual.u8, &next_hop_size,
-                          &next_hop_flags, (uint8_t *)dst,
-                          sizeof(ipv6_addr_t), 0) >= 0) &&
-        (next_hop_size == sizeof(ipv6_addr_t))) {
-        next_hop_ip = &next_hop_actual;
-    }
-
-#endif
-
-    if ((next_hop_ip == NULL)) {            /* no route to host */
-        if (iface == KERNEL_PID_UNDEF) {
-            /* ng_ipv6_netif_t doubles as prefix list */
-            iface = ng_ipv6_netif_find_by_prefix(&prefix, dst);
-        }
-        else {
-            /* ng_ipv6_netif_t doubles as prefix list */
-            prefix = ng_ipv6_netif_match_prefix(iface, dst);
-        }
 
-        if ((prefix != NULL) &&             /* prefix is on-link */
-            (ng_ipv6_netif_addr_get(prefix)->flags &
-             NG_IPV6_NETIF_ADDR_FLAGS_NDP_ON_LINK)) {
-            next_hop_ip = dst;
-#ifdef MODULE_FIB
-            /* We don't care if FIB is full, this is just for efficiency
-             * for later sends */
-            fib_add_entry(iface, (uint8_t *)dst, sizeof(ipv6_addr_t), 0,
-                          (uint8_t *)next_hop_ip, sizeof(ipv6_addr_t), 0,
-                          FIB_LIFETIME_NO_EXPIRE);
-#endif
-        }
-    }
-
-    if (next_hop_ip == NULL) {
-        next_hop_ip = ng_ndp_internal_default_router();
-#ifdef MODULE_FIB
-        /* We don't care if FIB is full, this is just for efficiency for later
-         * sends */
-        fib_add_entry(iface, (uint8_t *)dst, sizeof(ipv6_addr_t), 0,
-                      (uint8_t *)next_hop_ip, sizeof(ipv6_addr_t), 0,
-                      FIB_LIFETIME_NO_EXPIRE);
+    return ((fib_get_next_hop(iface, next_hop_ip->u8, &next_hop_size,
+                              &next_hop_flags, (uint8_t *)dst,
+                              sizeof(ipv6_addr_t), 0) >= 0) &&
+            (next_hop_size == sizeof(ipv6_addr_t)));
```

Comment:
Given your statement that lookup for link-local addresses is not necessary (I did not find any sources for this yet): shouldn't there be a check somewhere here that `dst` is a link-local address?